src/types/Typechecker/Typechecker.hs

```diff
@@ -1722,33 +1722,43 @@ instance Checkable Expr where
                                         ,stop = eStop
                                         ,step = eStep}
 
-    --  E |- rng : Range
-    --  E, x : int |- e : ty
-    -- --------------------------
-    --  E |- for x <- rng e : ty
-
-    --  E |- arr : [ty]
-    --  E, x : int |- e : ty
-    -- --------------------------
-    --  E |- for x <- arr e : ty
-    doTypecheck for@(For {name, step, src, body}) =
-        do stepTyped <- doTypecheck step
-           srcTyped  <- doTypecheck src
-           let srcType = AST.getType srcTyped
-
-           unless (isArrayType srcType || isRangeType srcType) $
-             pushError src $ NonIterableError srcType
-
-           let elementType = if isRangeType srcType
-                             then intType
-                             else getResultType srcType
-           bodyTyped <- typecheckBody elementType body
-           return $ setType unitType for{step = stepTyped
-                                        ,src  = srcTyped
-                                        ,body = bodyTyped}
-        where
-          addIteratorVariable ty = extendEnvironmentImmutable [(name, ty)]
-          typecheckBody ty = local (addIteratorVariable ty) . typecheck
+
+    -- JOY for-comprehension
+    -- returnType is unittye, is changed during the second typechecking when it has been desugaraed into a method call
+    doTypecheck for@(For {sources, body}) = do
+      sourcesTyped <- typeCheckSources sources
+      forVarList <- getForVarTypeList sources
+      bodyTyped <- typecheckBody forVarList body
+      return $ setType unitType for{sources = sourcesTyped,
+                                      body = bodyTyped}
+      where
+        typeCheckSources :: [ForSource] -> TypecheckM [ForSource]
+        typeCheckSources sourceList =  do
+          typedSources <- mapM typeCheckSource sourceList
+          return typedSources
+        typeCheckSource fors@(ForSource{forVarType, collection}) = do
+            collectionTyped <- doTypecheck collection
+            let collectionType = AST.getType collectionTyped
+            let forVarType = return $ getInnerType collectionType
+            return fors{forVarType = forVarType,
+                        collection = setType collectionType collectionTyped}
+
+        getForVarTypeList sourceList = mapM getForVarType sourceList
+        getForVarType ForSource{forVar, collection} = do
```

Comment:
I would add the typing to this function, I think it can serve as documentation giving that returning a tuple is not the most common thing for the compiler